<launch>
  <arg name="vending_machine_ar_id" />
  <arg name="predicted_robot_pose_topic" />
  <arg name="cmd_vel_mux_input"/>
  <arg name="spotted_markers_topic"/>
  <arg name="relative_target_pose_topic"/>
  <arg name="odom_topic"/>
  <arg name="robot_cmd_feed"/>
  <arg name="base_frame"/>
  <arg name="vm_interactor_action"      default="vm_interactor"/>
  <arg name="vel_smoother_param"        default="$(find yocs_vm_interactor)/params/vel_smoother_pose_controller.yaml"/>


  <node pkg="yocs_vm_interactor" name="vm_interactor" type="interactor.py"> 
    <param name="vending_machine_ar_id"  value="$(arg vending_machine_ar_id)"/>
    <remap from="vm_interactor" to="$(arg vm_interactor_action)"/>
    <remap from="cmd_vel" to="$(arg cmd_vel_mux_input)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
  </node>

  <node pkg="yocs_ar_pair_approach" type="approach.py" name="ar_pair_approach">
    <!--<remap from="ar_pair_approach/initialise"       to="$(arg predicted_robot_pose_topic)"/>-->
    <remap from="ar_pair_approach/enable"           to="enable_ar_pair_approach"/>
    <remap from="ar_pair_approach/result"           to="result_ar_pair_approach"/>
    <remap from="ar_pair_approach/cmd_vel"          to="$(arg cmd_vel_mux_input)"/>
    <remap from="ar_pair_approach/spotted_markers"  to="$(arg spotted_markers_topic)"/>
    <remap from="ar_pair_approach/relative_target_pose" to="$(arg relative_target_pose_topic)"/>
  </node>

  <node pkg="nodelet" type="nodelet" name="pose_controller_velocity_smoother" args="load yocs_velocity_smoother/VelocitySmootherNodelet /mobile_base_nodelet_manager">
    <rosparam file="$(arg vel_smoother_param)" command="load"/>
    <remap from="pose_controller_velocity_smoother/smooth_cmd_vel" to="$(arg cmd_vel_mux_input)"/>
    <remap from="pose_controller_velocity_smoother/robot_cmd_vel" to="$(arg robot_cmd_feed)"/>
    <remap from="pose_controller_velocity_smoother/odometry"  to="$(arg odom_topic)"/>
  </node>

  <node pkg="nodelet" type="nodelet" name="diff_drive_pose_controller" args="load yocs_diff_drive_pose_controller/DiffDrivePoseControllerNodelet /mobile_base_nodelet_manager">
    <param name="spin_rate" value="30"/>
    <param name="start_enabled" value="false"/>
    <param name="base_frame_name" value="$(arg base_frame)"/>
    <param name="v_min" value="0.01"/>
    <param name="v_max" value="0.10"/>
    <param name="k_1" value="1.0"/>
    <param name="k_2" value="10.0"/>
    <param name="beta" value="0.1"/>
    <param name="lambda" value="1.0"/>
    <param name="dist_thres" value="0.03"/>
    <param name="orient_thres" value="0.06"/>
    <param name="dist_eps" value="0.004"/>
    <param name="orient_eps" value="0.012"/>
<!--    <remap from="diff_drive_pose_controller/command_velocity" to="pose_controller_velocity_smoother/raw_cmd_vel"/>-->
    <remap from="diff_drive_pose_controller/command_velocity" to="$(arg cmd_vel_mux_input)"/>
    <remap from="diff_drive_pose_controller/enable"           to="ar_pair_approach/enable_approach_controller"/>
    <remap from="diff_drive_pose_controller/disable"          to="ar_pair_approach/disable_approach_controller"/>
    <remap from="diff_drive_pose_controller/pose_reached"     to="ar_pair_approach/approach_pose_reached"/>
  </node>
</launch>
